Automatic retry and recovery if IFCB connection is lost

figuernd opened the pull request:
Harvested from Oleander deployment. Used because IFCB is not always on, has to be shutdown sometimes to prevent overheating. We use the `ifcb/connected` topic to determine whether it is safe to start sending commands.

rgov commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -15,23 +16,30 @@
 from geometry_msgs.msg import Point
 from sensor_msgs.msg import CompressedImage
-from std_msgs.msg import ColorRGBA
+from std_msgs.msg import ColorRGBA, Bool
```
Not a big deal but I try to alphabetize the list (and all lists) rather than just append.

rgov commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -173,6 +211,81 @@ def on_triggerrois(roi_pub, mkr_pub, _, rois, *, timestamp=None):
+
+    while not rospy.is_shutdown():
+        try:
+            if connection_lost.is_set() or not ifcb_ready.is_set():
```
Reduce cyclomatic complexity by inverting this condition:

```python
if ifcb_ready.is_set() and not connection_lost.is_set():
    time.sleep(5)  # or is there a rospy function we're supposed to use?
    continue
```

figuernd replied:
Done in 108c59a10df6668408b1acabc9f85552a641d62e

rgov commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -173,6 +211,81 @@ def on_triggerrois(roi_pub, mkr_pub, _, rois, *, timestamp=None):
+
+                # Create the client
+                try:
+                    client = IFCBClient(
+                        f'ws://{rospy.get_param("~address")}'\
+                            f':{rospy.get_param("~port", 8092)}/ifcbHub',
+                        rospy.get_param('~serial'),
+                    )
```
Can we keep a single `IFCBClient` instance and just reconnect it?

figuernd replied:
Okay, I don't really want a state where ifcb messages queue up if the IFCB is offline though.

Would you mind if we revisit this another time? I get that pyifcbclient could be optimized for our purpose here but that feels like scope creep.

rgov commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -15,23 +16,30 @@
 
 
 ifcb_ready = threading.Event()
+connection_lost = threading.Event()
```
Also thinking that the onus will now be on the service caller to pay attention to whether the IFCB has been rebooted. That's complexity we haven't thought about before.

figuernd commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -173,6 +211,81 @@ def on_triggerrois(roi_pub, mkr_pub, _, rois, *, timestamp=None):
+            # Connection monitoring - check every few seconds
+            time.sleep(5)
+
+        except Exception:
```
Added 108c59a10df6668408b1acabc9f85552a641d62e

rgov commented on src/ifcb/src/ifcb/node.py:
```diff
@@ -80,30 +82,33 @@ def do_runroutine(pub, req):
 
     # Wait for the IFCB to be ready; this prevents sending a command too early
-    ifcb_ready.wait()
+    ready = ifcb_ready.wait(5.0)
+    if not ready:
+        return False
```
Minor not: I wouldn't use an intermediate variable here:

```python
if not ifcb_ready.wait(5.0):
    # log something, it's unusual for the ifcb to time out
    return False
```

rgov reviewed (commented):
Looking better, yeah.

rgov reviewed (approved):
Looks good to me. How do we test it?

figuernd commented:
Going to test it on the Oleander when it gets back to Bermuda

figuernd commented:
@rgov if you could take one more look at https://github.com/WHOIGit/PhytO-ARM/pull/91/files#diff-030989f224bb7bf9bb595322b36f9d3387e57f18a27aca806431bd1a2f42a09c

I've refactored it to use a single client and simplified the number of status variables down to just `ifcb_ready`.

Likewise I've added a flag for awaiting the IFCB in the event it is unavailable: https://github.com/WHOIGit/PhytO-ARM/pull/91/files#diff-4cad7f49b3bfae8e253237741011e61be2965f1e6edfa36a46fc97bd9d70bbf4

This isn't meant to make PhytO-ARM immune to IFCB failures, but lets the arm stay in a holding pattern when we have planned downtime.

rgov reviewed (commented):
Partial review, I'm not able to read through it all right now.

> Note that the initial connect() method for signalrcore does not seem to utilize the same unlimited re-attempts that reconnect does

Is that why you still have reconnect logic in this? Would it be desirable for PhytO-ARM to refuse to launch if IFCBacquire is not reachable at all?
